### ISSUE TITLE
Nodes stop receiving UDP Packets after a while

### ISSUE THREAD

**Issue by Diggensag** [intrig-unicamp/mininet-wifi#249]
I am starting to really feel sorry for opening so many issues.

It could be possible that my code is wrong, so i created a fresh fork, with a stripped down version of my project to show the issue.

Thing is, I can run the same code for a long time (30 min) on real hardware and I have no problems at all.
When I start my emulation in mininet wifi, after one or two minutes my sta*X* stops receiving UDP-packets - but I can see them via Wireshark on hwsim0.

In my example I have 5 stations: the "car" stops at the fifth station and the car keeps telling it's position via ipv6 multicast, after some (sporadic, not a certain) time the node stops receiving these UDP packets.

Have you ever experienced something like that?

**Comment by ramonfontes** [comment]
No, I've never experienced something like that.

I have some questions:
1. How are the nodes communicating with each other? 
2. Is it a non-infrastructure or infrastructure scenario?
3. Who is the destination? 
4. Is OpenFlow being used? If so, do you have any comments about the flows?
5. Why are you using hwsim0? What do you think about creating the monitor interface manually?

Well, I have few information and many questions. It's really difficult to make some assumption about this.

**Comment by Diggensag** [comment]
1. via local link multicast IPv6
``` 
def runMulticastListener(interfaceV6, nodelogger):
    multicastaddress = "ff02::1" + "%" + interfaceV6
    hadd = socket.getaddrinfo(multicastaddress, 8080, socket.AF_INET6, socket.SOCK_DGRAM)
    haddr = socket.getaddrinfo(multicastaddress, 8080, socket.AF_INET6, socket.SOCK_DGRAM)[0][-1]
    
    # Initialise socket for IPv6 datagrams
    sock = socket.socket(socket.AF_INET6, socket.SOCK_DGRAM, socket.IPPROTO_UDP)

    # Allows address to be reused
    sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)

    # Binds to interface
    sock.bind(haddr)

    # Allow messages from this socket to loop back for development
    sock.setsockopt(socket.IPPROTO_IPV6, socket.IPV6_MULTICAST_LOOP, True)

    # Construct message for joining multicast group
    mreq = struct.pack("16s15s".encode('utf-8'), socket.inet_pton(socket.AF_INET6, "ff02::1"), (chr(0) * 16).encode('utf-8'))
    sock.setsockopt(socket.IPPROTO_IPV6, socket.IPV6_JOIN_GROUP, mreq)

    run = True
    try:
        while run==True:
            data, addr = sock.recvfrom(1024)
            nodelogger.debug("\nGot Message: " + str(data) +" | from: " + str(addr))
...

```

2. non infrastructure
3. all neighbors
4. not sure, I think no - I have not defined any rules
5. I need to get a overview over the complete traffic. In the meantime I hat the same idea and opened a wireshark on sta5-wlan and saw the packets there, too. So I am now further investigating whether my UDP Implementation is faulty (I don't get why it's working in 'reallife').

**Comment by ramonfontes** [comment]
The code itself doesn't help. You would help if you provided more information about the problem. There is no description of the topology; mobility pattern; routing protocol, etc; Sorry, but I cannot help you with so poor information.

**Comment by Diggensag** [comment]
It`s a small 7 node setup (6 non movable stations + 1 station used as "car"), every node is in a adhoc-network. I am using wmediumd with interference and logDistance
```
    net = Mininet_wifi(link=wmediumd, wmediumd_mode=interference)

    info("*** Creating nodes\n")
    nodes = {}
    
    nodes['sta1'] = net.addStation('sta1', position='10,10,0')
    ...
    nodes['sta5'] = net.addStation('sta5', position='90,10,0')

    car1 = net.addStation('car1', position='10.0,20.0,0.0', range=30)
    arbiter1 = net.addStation('arbiter1', position='50.0,12.0,0.0', range=50)

*snip*
    net.setPropagationModel(model="logDistance", exp=4)
*snip*

    for key in nodes.keys():
        net.addLink(nodes[key], cls=adhoc, ssid=wifi_ssid, mode=wifi_mode, channel=wifi_channel, ht_cap='HT40+')

    net.addLink(car1, cls=adhoc, ssid='6mesh', mode='g', channel=11, ht_cap='HT40+')
    net.addLink(arbiter1, cls=adhoc, ssid='6mesh', mode='g', channel=11, ht_cap='HT40+')
```
The mobility pattern is, that only one station has static mobility moving from left to right:
```
info("*** Setting up static mobility\n") 
    net.startMobility(time=50, repetitions=1)
    net.mobility(car1, 'start', time=51, position='10.0,15.0,0.0')
    net.mobility(car1, 'stop', time=180, position='90.0,15.0,0.0')
    net.stopMobility(time=181)
```
There is no real routing, everything is send link-local to all neighbors via the code.
The car-node is sending it`s own position every second, all static nodes capture these packets and write them to a log file. After a uncertain time, let's say 2 minutes or so, I experience that the programm stops receiving the packets from the car but i still can see the UDP traffic incoming on staX-wlan0.

It's late over here. Tomorrow I will try to reduce it further more. I will test it stripped down to two nodes, one sending, one receiving, to see if I still can reproduce this and compare it to two notebooks running the same code via real wifi.

**Comment by ramonfontes** [comment]
So, there is no routing protocol being used in this scenario and it works with physical nodes? Unless you tell me how this is possible, I cannot view this as an issue. Could you show me this working with the simplest topology? Please refer to `/examples/adhoc.py`.

**Comment by Diggensag** [comment]
Sorry for the late reply, it was 3 AM this morning and the wifi was stuck, now finally online again after some hours of sleep. I can give you  a minimal example later today if you still want it then. Think it's unnecessary now :-)

I went on with a minimal setup and found a interesting pitfall. So this may not be an issue towards the network itself, but it turns out, that mininet-wifi (or also only mininet) has kind of a maximum buffer for program-output. Maybe this is something to note for future reference.

So me starting my program via sta1.cmd("myprogram.py &") and having it acting like a server is not the problem. The problem seems to be a maximum console buffer space for each program run in a cmd-shell. So my programm has some vector calculations etc. and i was printing them out for debug purposes, also for me to see (on the reallife nodes) what is going on and what the server is doing. So filling this buffer the programm stops running / is stuck at the point where it cannot print more characters to the console.

This just my assumption, but it would perfectly explain why my programs went well on realworld notebooks / raspberryPIs -  they could print everything out to the bash. I was pointed to this when I tried to prepare my minimal setup: I started my program on one host via xterm then and it did not stop receiving packets.

Finally i tossed out all "print" commands and replaced them with logging commands and it is working now till the end.

I am so happy :-D
Took me three days to find this.
But in the end you are right, again it's not an issue to the network or mininet-wifi itself. But it is a hard pitfall, that I want to document, so nobody makes my mistakes a second time.
